src/prommis/uky/costing/ree_plant_capcost.py

```diff
@@ -2675,3 +2736,672 @@ def costing_upper_bound_eq(c, p):
+                            to_units=pyunits.dimensionless,
+                        ),
+                        len(capital_expenditure_percentages),
+                    )
+                ),
+                to_units=cost_units,
+            )
+
+        @b.Constraint()
+        def pv_revenue_constraint(c):
+            # PV_Revenue = REVENUE * [ P/A_OPEX+CAPEX_periods - P/A_CAPEX_period ]
+
+            return c.pv_revenue == pyunits.convert(
+                REVENUE
+                * (
+                    series_present_worth_factor(
+                        pyunits.convert(
+                            c.discount_percentage, to_units=pyunits.dimensionless
+                        ),
+                        pyunits.convert(
+                            c.revenue_inflation_percentage,
+                            to_units=pyunits.dimensionless,
+                        ),
+                        c.plant_lifetime / pyunits.year
+                        + len(capital_expenditure_percentages),
+                    )
+                    - series_present_worth_factor(
+                        pyunits.convert(
+                            c.discount_percentage, to_units=pyunits.dimensionless
+                        ),
+                        pyunits.convert(
+                            c.revenue_inflation_percentage,
+                            to_units=pyunits.dimensionless,
+                        ),
+                        len(capital_expenditure_percentages),
+                    )
+                ),
+                to_units=cost_units,
+            )
+
+        if royalty_expression is None:
+            # PV_Royalties = - %royalty_charge_of_revenue * REVENUE
+
+            @b.Constraint()
+            def pv_royalties_constraint(c):
+
+                return c.pv_royalties == -pyunits.convert(
+                    pyunits.convert(
+                        c.royalty_charge_percentage_of_revenue,
+                        to_units=pyunits.dimensionless,
+                    )
+                    * c.pv_revenue,
+                    to_units=cost_units,
+                )
+
+        else:
+
+            @b.Constraint()
+            def pv_royalties_constraint(c):
+
+                return c.pv_royalties == -pyunits.convert(
+                    royalty_expression, to_units=cost_units
+                )
+
+        if debt_expression is None:
+
+            @b.Constraint()
+            def loan_debt_constraint(c):
+                # Debt  = - %debt_charge_of_CAPEX * CAPEX
+
+                return c.loan_debt == pyunits.convert(
+                    pyunits.convert(
+                        c.debt_percentage_of_CAPEX, to_units=pyunits.dimensionless
+                    )
+                    * CAPEX,
+                    to_units=cost_units,
+                )
+
+        else:
+
+            @b.Constraint()
+            def loan_debt_constraint(c):
+
+                return c.loan_debt == pyunits.convert(
```

Comment:
Do you need the constraint in these cases? You are adding a lot of simple linear equality constraints to the model which is adding a lot of unnecessary burden on the solver (and writer).

I would suggest for these cases making `c.load_debt` a `Reference` based on the argument provided (noting you will need to move the unit conversion elsewhere).

Reply:
Similarly, you might want to break these sections out as separate sub-methods to make it easier to test them in isolation (so that you can set up a simple Block with the things you need for each branch of the `if/else` and make sure the variables/references and constraints are built correctly).